accord-core/src/main/java/accord/local/durability/DurabilityQueue.java

```diff
@@ -73,325 +86,702 @@ public class DurabilityQueue
+                        if (fail instanceof SyncPointErased || fail instanceof TopologyRetiredException)
+                        {
+                            if (request == null) abandon(pending, false, false);
+                            else restart(pending);
+                            return;
+                        }
+                    }
+
+                    if (fail instanceof Exhausted || success != null)
+                    {
+                        Collection<Node.Id> failedNodes = success != null ? success.min.excluding : ((Exhausted)fail).failedNodes();
+                        Ranges failedRanges = success != null ? ranges : ((Exhausted)fail).failedRanges();
+                        boolean log = failedNodes == null;
+                        if (!log)
+                        {
+                            Set<Node.Id> unlogged = new HashSet<>(failedNodes);
+                            for (Collection<Node.Id> logged : WARNINGS_LOGGED.tailMap(System.nanoTime() - MINUTES.toNanos(EXHAUSTED_LOG_INTERVAL_MINUTES)).values())
+                                unlogged.removeAll(logged);
+                            log = !unlogged.isEmpty();
+                        }
+                        if (log)
+                        {
+                            logger.info("{}: Incomplete durability for {}{}. {} were unsuccessful.", txnId, failedRanges, requestor, failedNodes == null ? "some nodes" : failedNodes);
+                            WARNINGS_LOGGED.headMap(System.nanoTime() - MINUTES.toNanos(EXHAUSTED_LOG_INTERVAL_MINUTES)).clear();
+                            WARNINGS_LOGGED.put(System.nanoTime(), failedNodes == null ? Collections.emptyList() : failedNodes);
+                        }
+                    }
+                    else
+                    {
+                        if (fail instanceof Timeout) logger.info("{}: Timeout awaiting durability for {}{}", txnId, ranges, requestor, fail);
+                        else if (fail != null) logger.info("{}: Failed awaiting durability for {}{}; will retry", txnId, ranges, requestor, fail);
+                    }
+
+                    if (achieved.intersects(ranges))
+                    {
+                        unregister(pending);
+                        pending.syncPoint = pending.syncPoint.without(achieved);
+                        pending.updateByRange();
+                        register(pending);
+                    }
 
-                retry(exclusiveSyncPoint, request, attempt + 1);
+                    retry(pending);
+                }
             }
-            else if (request == null)
+            finally
             {
-                // if request != null, the request will log the necessary information
-                logger.debug("{}: Successfully achieved durability for {}.", txnId, ranges);
+                synchronized (this)
```

Comment:
Maybe that's just me, but could you add a DurabilityQueue qualifier to this, since we're in lambda?

Reply:
lambdas don't have a `this` qualifier, but I can add this if it helps